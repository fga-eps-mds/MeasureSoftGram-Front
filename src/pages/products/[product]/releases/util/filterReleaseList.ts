import { CompareGoalAccomplished } from '@customTypes/product';

function filterReleaseList(
  releases: CompareGoalAccomplished[],
  name: string,
  startDate: string,
  endDate: string
): CompareGoalAccomplished[] {
  return releases.filter((release) => {
    const endDateObject = new Date(endDate);
    endDateObject.setUTCHours(23, 59, 59);
    const hasName = !name || release.release_name.toLowerCase().includes(name.toLowerCase());
    const isAfterStartDate = !startDate || new Date(release.start_at) >= new Date(startDate);
    const isBeforeEndDate = !endDate || new Date(release.end_at) <= endDateObject;
    return isAfterStartDate && hasName && isBeforeEndDate;
  });
}

export default filterReleaseList;
