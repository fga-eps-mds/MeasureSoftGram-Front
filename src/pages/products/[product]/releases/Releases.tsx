import React, { useMemo, useState } from 'react';
import Head from 'next/head';

import { Box, Container, Pagination, Skeleton, Stack, TextField, Typography } from '@mui/material';

import { NextPageWithLayout } from '@pages/_app.next';

import getLayout from '@components/Layout';

import { productQuery } from '@services/product';
import { useProductContext } from '@contexts/ProductProvider';
import { useOrganizationContext } from '@contexts/OrganizationProvider';
import { useRequest } from '@hooks/useRequest';
import { CompareGoalAccomplished } from '@customTypes/product';
import SearchButton from '@components/SearchButton';
import ReleasesTable from '../components/ReleasesList/ReleasesTable';
import filterReleaseList from './util/filterReleaseList';

const RELEASES_PER_PAGE = 10;

const Releases: NextPageWithLayout = () => {
  const [startDate, setStartDate] = useState<string>('');
  const [endDate, setEndDate] = useState<string>('');
  const [name, setName] = useState<string>('');
  const [page, setPage] = useState<number>(1);
  const { currentProduct } = useProductContext();
  const { currentOrganization } = useOrganizationContext();
  const { data: releaseList, isLoading } = useRequest<CompareGoalAccomplished[]>(
    productQuery.getReleaseList(currentOrganization?.id as string, currentProduct?.id as string)
  );

  const filteredReleaseList = useMemo(() => filterReleaseList(releaseList ?? [], name, startDate, endDate), [releaseList, name, startDate, endDate])

  return isLoading ? (<Skeleton />) :
    <>
      <Head>
        <title> Releases </title>
      </Head>

      <Container>
        <Box display="flex" flexDirection="column">
          <Box display="flex" gap="1rem" marginTop="40px" marginBottom="36px">
            <Box display="flex" alignItems="center">
              <Typography variant="h4" marginRight="10px" color="#33568E" fontWeight="500">
                Releases
              </Typography>
            </Box>
          </Box>
        </Box>
        <Box display="flex" flexDirection="column" padding="20px"
          style={{ backgroundColor: 'white', border: '1px solid #113d4c80', borderRadius: '10px' }}>
          <Box display="flex" justifyContent="space-evenly">
            <Box>
              <Typography color="#538BA3">
                In√≠cio da release
              </Typography>
              <TextField
                type="date"
                required
                value={startDate}
                onChange={(e) => setStartDate(e.target.value)}
                inputProps={{
                  'data-testid': 'inicio-release'
                }}
                size='small'
              />
            </Box>

            <Box>
              <Typography color="#538BA3">
                Fim da release
              </Typography>
              <TextField
                type="date"
                required
                size='small'
                value={endDate}
                onChange={(e) => setEndDate(e.target.value)}
                inputProps={{
                  'data-testid': 'fim-release'
                }}
              />
            </Box>
            <Box marginTop="16px">
              <SearchButton
                label='Buscar releases'
                placeHolder='Insira o nome da release'
                onInput={(e) => setName(e.target.value)}
              />
            </Box>
          </Box>
          <ReleasesTable releaseList={(filteredReleaseList.slice((page - 1) * RELEASES_PER_PAGE, page * RELEASES_PER_PAGE))} />
          <Stack spacing={2} sx={{ mt: 2 }} justifyContent="center" alignItems="center">
            <Pagination
              count={Math.ceil(filteredReleaseList.length / RELEASES_PER_PAGE)}
              onChange={(e, newPage) => { setPage(newPage) }}
              color='primary'
              size='large'
              shape="rounded"
            />
          </Stack>
        </Box>
      </Container >
    </>
};

Releases.getLayout = getLayout;

export default Releases;
