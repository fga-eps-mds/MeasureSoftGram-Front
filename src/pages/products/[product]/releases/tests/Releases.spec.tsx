import { act, fireEvent, render } from "@testing-library/react";
import { useRequest } from "@hooks/useRequest";
import Releases from "../Releases";
import filterReleaseList from "../util/filterReleaseList";

jest.mock('@contexts/ProductProvider', () => ({
  useProductContext: jest.fn(() => ({
    currentProduct: {
      id: '1',
      name: 'MeasureSoftGram'
    }
  }))
}));

jest.mock('@contexts/OrganizationProvider', () => ({
  useOrganizationContext: jest.fn(() => ({
    currentOrganization: {
      id: '1'
    }
  }))
}));

jest.mock('@hooks/useRequest', () => ({
  useRequest: jest.fn(() => ({
    data: [],
    isLoading: false
  }))
}));

jest.mock('../util/filterReleaseList', () => jest.fn(() => []));

interface MockPaginationProps {
  // eslint-disable-next-line no-unused-vars
  onChange: (_event: React.ChangeEvent<HTMLSelectElement>) => void;
}

const MockPagination = ({ onChange }: MockPaginationProps) => (
  <select data-testid="select-page" onChange={onChange}>
    <option value="1">1</option>
    <option value="2">2</option>
  </select>
)

jest.mock('@mui/material', () => ({
  ...jest.requireActual('@mui/material'),
  Pagination: (props: any) => <MockPagination {...props} />
}));

describe('<Releases />', () => {
  it('Deve corresponder ao snapshot', async () => {
    const { asFragment } = render(<Releases />);

    expect(asFragment()).toMatchSnapshot();
  })

  it('Deve filtrar as releases pelo nome, data de início e data de fim', async () => {
    const { getByTestId, getByRole } = render(<Releases />);

    const startDate = getByTestId('inicio-release');
    const endDate = getByTestId('fim-release');
    const searchButton = getByRole('textbox', { name: /Buscar releases/i });

    await act(async () => {
      fireEvent.change(startDate, { target: { value: '2021-01-01' } });
      fireEvent.change(endDate, { target: { value: '2021-01-02' } });
      fireEvent.input(searchButton, { target: { value: 'MeasureSoftGram' } });
    });

    expect(filterReleaseList).toHaveBeenCalledWith([], 'MeasureSoftGram', '2021-01-01', '2021-01-02');
  })

  it('Deve mostrar ao menos uma opção de página', async () => {
    const { getByTestId } = render(<Releases />);

    const pageSelect = getByTestId('select-page');

    fireEvent.change(pageSelect, { target: { value: '1' } });
    expect(pageSelect).toBeDefined();
  })

  it('Deve chamar o filterReleaseList com array vazia quando o releaseList for undefined', async () => {
    (useRequest as jest.Mock).mockImplementationOnce(() => ({
      data: undefined,
      isLoading: false
    }));

    render(<Releases />);

    expect(filterReleaseList).toHaveBeenCalledWith([], '', '', '');
  })

  it('Deve mostrar o Skeleton enquanto o isLoading for true', async () => {
    (useRequest as jest.Mock).mockImplementationOnce(() => ({
      data: undefined,
      isLoading: true
    }));

    const { queryByText } = render(<Releases />);
    expect(queryByText('Nome')).toBeNull();
  })
});
