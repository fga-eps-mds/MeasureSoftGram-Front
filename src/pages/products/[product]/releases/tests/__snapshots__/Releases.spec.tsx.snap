// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Releases /> Deve corresponder ao snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-j7qwjs"
    >
      <div
        class="MuiBox-root css-1k2nw5y"
      >
        <div
          class="MuiBox-root css-70qvj9"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-1f3qdbj-MuiTypography-root"
          >
            Releases
          </h4>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1k9wtw"
      style="background-color: white; border: 1px solid #113d4c80; border-radius: 10px;"
    >
      <div
        class="MuiBox-root css-102lqhp"
      >
        <div
          class="MuiBox-root css-0"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-3lak7q-MuiTypography-root"
          >
            Início da release
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="inicio-release"
                id=":r0:"
                required=""
                type="date"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-3lak7q-MuiTypography-root"
          >
            Fim da release
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="fim-release"
                id=":r1:"
                required=""
                type="date"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1yuhvjn"
        >
          <button
            aria-label="search"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              style="fill: #113d4c;"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiFormControl-root MuiTextField-root text css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="search-bar"
              id="search-bar-label"
            >
              Buscar releases
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id="search-bar"
                placeholder="Insira o nome da release"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Buscar releases
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
      >
        <table
          class="MuiTable-root css-rqglhn-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Nome
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Início da release
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Fim da release
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          />
        </table>
      </div>
      <div
        class="css-1pimw9g-MuiStack-root"
      >
        <select
          data-testid="select-page"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
